File: go_files/ex15.go
package main

import (
	"fmt"
)

type SaldoInsuficienteError struct {
	saldo  float64
	valor  float64
}

func (e *SaldoInsuficienteError) Error() string {
	return fmt.Sprintf("saldo insuficiente: disponível R$%.2f\nSaque solicitado R$%.2f", e.saldo ,e.valor)
}

type ContaBancaria struct {
	titular string
	saldo   float64
}

func (c *ContaBancaria) Depositar(valor float64) {
	c.saldo += valor
}

func (c *ContaBancaria) Sacar(valor float64) error {
	if valor > c.saldo {
		return &SaldoInsuficienteError{saldo: c.saldo, valor: valor}
	}
	c.saldo -= valor
	return nil
}

func main() {

	conta := ContaBancaria{titular: "Carlos", saldo: 90.00}

	if err := conta.Sacar(150.00); err != nil {
		fmt.Println("Erro:", err)
	} else {
		fmt.Println("Saque realizado com sucesso!")
	}

	conta.Depositar(100.00)

	if err := conta.Sacar(50.00); err != nil {
		fmt.Println("Erro:", err)
	} else {
		fmt.Println("Saque realizado com sucesso!")
	}
}

File: pyton_files/ex15.py
class SaldoInsuficienteException(Exception):
    def __init__(self, saldo, valor_saque):
        super().__init__(f"Saque de valor {valor_saque} não liberado. Saldo disponível para saque é de: {saldo}")

class ContaBancaria:
    def __init__(self, saldo):
        self.saldo = saldo

    def sacar(self, valor):
        if valor > self.saldo:
            raise SaldoInsuficienteException(self.saldo, valor)
        self.saldo -= valor
        print(f"Saque realizado com sucesso. Saldo atual é de : {self.saldo}")

# Exemplo de uso
conta = ContaBancaria(1000)

try:
    conta.sacar(1500)
except SaldoInsuficienteException as e:
    print(e)  